File: python/maximum_product_of_splitted_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxProduct(self, root: TreeNode) -> int:
        self.sums = []
        self.max_product = 1
        #traverse the tree and get the sum of every subtree and add them to self.sums
        #because we're using post-order traversal, the total sum will be returned last
        total_sum = self.getSums(root)
        #for each subtree, calculate the product of the two trees formed by excizing the subtree from the tree
        for current_sum in self.sums:
            current_product = current_sum * (total_sum - current_sum)
            self.max_product = max(self.max_product, current_product)
        return self.max_product % 1000000007
        
    def getSums(self, root) -> int:
        #recursive base case
        if root is None:
            return 0
        #use post-order traversal
        left = self.getSums(root.left)
        right = self.getSums(root.right)
        current_sum = left + right + root.val
        self.sums.append(current_sum)
        return current_sum

File: java/maximumProductOfSplittedBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.math.BigInteger;

class Solution {
    List<Integer> sums = new ArrayList();
    
    public int maxProduct(TreeNode root) {
        int totalSum = getSums(root);
        BigInteger maxProduct = BigInteger.valueOf(1);
        for (int s : sums) {
            BigInteger prod = BigInteger.valueOf(s * (totalSum - s));
            if (prod.compareTo(maxProduct) == 1) {
                maxProduct = prod;
            }
        }
        return (maxProduct.mod(BigInteger.valueOf(1000000007))).intValue();
    }
    
    private int getSums(TreeNode root) {
        //recursive base case
        if (root == null) {
            return 0;
        }
        int currentSum = getSums(root.left) + getSums(root.right) + root.val;
        sums.add(currentSum);
        return currentSum;
    }
}